pageserver/src/tenant/timeline.rs

```diff
@@ -2801,7 +2818,18 @@ impl Timeline {
         let mut timeline = self;
 
         let mut read_count = scopeguard::guard(0, |cnt| {
-            crate::metrics::READ_NUM_LAYERS_VISITED.observe(cnt as f64)
+            crate::metrics::READ_NUM_LAYERS_VISITED.observe(cnt as f64);
+            if cnt >= Self::LAYERS_VISITED_WARN_THRESHOLD {
+                use utils::rate_limit::RateLimit;
+                static LOGGED: Lazy<Mutex<RateLimit>> =
+                    Lazy::new(|| Mutex::new(RateLimit::new(Duration::from_secs(10))));
+                let mut rate_limit = LOGGED.lock().unwrap();
+                let rate_limited = rate_limit.rate_limited();
+                rate_limit.call(|| {
+                    warn!("Too many layers visited by non-vectored read: {} >= {}; rate limited {} occurences",
+                          cnt, Self::LAYERS_VISITED_WARN_THRESHOLD, rate_limited);
```

Comment:
Could be more structured:

```suggestion
                    warn!(visited=cnt, threshold=Self::LAYERS_VISITED_WARN_THRESHOLD, suppressed, "Too many layers visited by non-vectored read");
```

These will be logged within a span which is per-tenant and per-timeline. My initial instinct would be that this tenant had X other messages rate limited or supresssed, but perhaps we'll remember those are from unrelated tenants.